### DISCUSSION TITLE
如何动态指定关联表的别名？

### DISCUSSION THREAD

**Discussion by Mason-3** [dotnetcore/FreeSql#1458] [Q&A] [answered]
.InnerJoin<ProjectStrPropertyEntity>((t, t1) => t.TaskId == t1.TaskId && t1.PropertyCode == pDetail.PropertyCode && t1.PropertyValue == pDetail.PropertyValue.ToString());如何动态指定t1的别名？因为这段join代码在一个foreach中

**Comment by densen2014** [comment]
仅供参考

```
var tempTable="myTable";
                fsql.Select<ProductsBase, ProductsImport>().AsTable((type, oldname) => {

if (type == typeof(ProductsImport) then return tempTable;
return oldname;

)
                    .InnerJoin(a => a.BarCode == a.BarCode)
                    .ToList();

```

**Reply by Mason-3** [reply]
Hi,我这段代码需要动态Join表，代码Demo:
var propertySelect = _freeSql.Select<ProjectStrPropertyEntity>();

            if (search.PropertyList.Any())
            {
                foreach (var pDetail in search.PropertyList)
                {                                    
                    switch (pDetail.PropertyType)
                    {
                        case PropertyTypeConstant.String:
                            propertySelect = propertySelect
                                    .InnerJoin<ProjectStrPropertyEntity>((t, t1) => t.TaskId == t1.TaskId && t1.PropertyCode == pDetail.PropertyCode && t1.PropertyValue == pDetail.PropertyValue.ToString());
                            break;
                        case PropertyTypeConstant.Int:
                            if (int.TryParse(pDetail.PropertyValue, out int intValue))
                            {
                                propertySelect = propertySelect
                                    .InnerJoin<ProjectIntPropertyEntity>((t, t1) => t.TaskId == t1.TaskId && t1.PropertyCode == pDetail.PropertyCode && t1.PropertyValue == intValue);
                            }
                            break;
                        case PropertyTypeConstant.Boolean:
                            if (bool.TryParse(pDetail.PropertyValue, out bool boolValue))
                            {
                                propertySelect = propertySelect
                                    .InnerJoin<ProjectBoolPropertyEntity>((t, t1) => t.TaskId == t1.TaskId && t1.PropertyCode == pDetail.PropertyCode && t1.PropertyValue == boolValue);
                            }
                            break;
                        case PropertyTypeConstant.DateTime:
                            if (DateTime.TryParse(pDetail.PropertyValue, out DateTime dtValue))
                            {
                                propertySelect = propertySelect
                                    .InnerJoin<ProjectDateTimePropertyEntity>((t, t1) => t.TaskId == t1.TaskId && t1.PropertyCode == pDetail.PropertyCode && t1.PropertyValue == dtValue);
                            }
                            break;
                        default:
                            continue;
                    }
                }               
            },
这段代码问题是最后join表的别名都是t1，我在想怎么修改这个t1，可以自己自定义吗，谢谢您的回复

**Comment by 2881099** [comment, marked as answer]
动态 Join，使用导航属性会方便一些。

有导航属性，返回的依然是 ISelect\<T\>，排序、ToList 都比较方便，如下：

```c#
class T
{
    public T1 t1 { get; set; }
}

fsql.Select<T>()
   .InnerJoin(a => a.t1.TaskId == a.TaskId && ...)
   .OrderBy(a => a.t1.code)
  .ToList(a => new { item = a, t2 = a.t1);
```

没有导航属性，返回的是 ISelect\<T\> 或者 ISelect\<T, T1\>

```c#
fsql.Select<T>()
    .InnerJoin<T1>((a, t1) => ...)
    .OrderBy(a => ...) //此时取 t1 较麻烦

fsql.Select<T, T1>() //泛型数量的固定，不如上面的方便
    .InnerJoin((a, t1) => ...)
    .ToList((a, t1) => new { item = a, t1 = t1 });
```
